harness/determined/common/api/certs.py

```diff
@@ -0,0 +1,228 @@
+            # (see https://docs.python.org/3/library/tempfile.html#tempfile.NamedTemporaryFile)
+            fd, combined_path = tempfile.mkstemp(prefix="det-master-cert-")
+            atexit.register(os.unlink, combined_path)
+
+            with open(fd, "wb") as out:
+                with open(certifi.where(), "rb") as base_certs:
+                    out.write(base_certs.read())
+                out.write(b"\n")
+                out.write(cert_pem.encode("utf8"))
+
+            self._bundle = combined_path
+
+    @property
+    def bundle(self) -> Union[None, str, bool]:
+        """
+        The path to a file containing an SSL certificate to trust specifically for the master, if
+        any, or False to disable cert verification entirely. If set to a path, it should always be a
+        temporary file that we own and can delete.
+        """
+        return self._bundle
+
+    @property
+    def name(self) -> Optional[str]:
+        """
+        The name we use to verify the master certificate.
+        """
+        return self._name
+
+
+cli_cert = None  # type: Optional[Cert]
+
+
+class CertStore:
+    """
+    CertStore represents a persistent file-based record of certificates, each associated with a
+    particular master url.
+    """
+
+    def __init__(self, path: pathlib.Path) -> None:
+        self.path = path
+
+    @contextlib.contextmanager
+    def _persistent_store(self) -> Iterator[Dict["str", "str"]]:
+        """
+        Yields the appropriate store that can be modified, and the modified result will be written
+        back to file.
+        """
+        self.path.parent.mkdir(mode=0o700, parents=True, exist_ok=True)
+        # Decide on paths for a lock file and a temp files (during writing).
+        temp = pathlib.Path(str(self.path) + ".temp")
+        lock = pathlib.Path(str(self.path) + ".lock")
+
+        with filelock.FileLock(lock):
+            store = _load_cert_store(self.path)
+
+            # No need for try/finally, because we don't update the file after failures.
+            yield store
+
+            with temp.open("w") as f:
+                json.dump(store, f, indent=4, sort_keys=True)
+            temp.rename(self.path)
+
+    def get_cert(self, url: str) -> Optional[str]:
+        """
+        get_cert returns the contents of a cert (if any) that has been associated with the given
+        url.
+        """
+        if not self.path.exists():
+            return None
+        # Technically this doesn't have to be modfiable, but it is unlikely to matter.
+        with self._persistent_store() as store:
+            return store.get(url)
+
+    def set_cert(self, url: str, cert_pem: str) -> None:
+        with self._persistent_store() as store:
+            store[url] = cert_pem
+
+    def delete_cert(self, url: str) -> None:
+        with self._persistent_store() as store:
+            if url in store:
+                del store[url]
+
+
+def _load_cert_store(path: pathlib.Path) -> Dict[str, str]:
```

Comment:
this should return `CertStore` since it is now defined as a separate class?

Reply:
Renamed this and attached it as a method to the CertStore object